public/app/features/library-panels/components/PanelLibraryOptionsGroup/PanelLibraryOptionsGroup.tsx

```diff
@@ -11,29 +12,47 @@ import { LibraryPanelsView } from '../LibraryPanelsView/LibraryPanelsView';
 import { PanelQueriesChangedEvent } from 'app/types/events';
 import { LibraryPanelDTO } from '../../types';
 import { toPanelModelLibraryPanel } from '../../utils';
+import { connect, MapDispatchToProps } from 'react-redux';
+import { changePanelPlugin } from 'app/features/dashboard/state/actions';
 
 interface Props {
   panel: PanelModel;
   dashboard: DashboardModel;
+  onChange: () => void;
+}
+interface DispatchProps {
+  changePanelPlugin: typeof changePanelPlugin;
 }
 
-export const PanelLibraryOptionsGroup: React.FC<Props> = ({ panel, dashboard }) => {
+export const PanelLibraryOptionsGroupUnconnected: React.FC<Props & DispatchProps> = ({
+  panel,
+  dashboard,
+  onChange,
+  changePanelPlugin,
+}) => {
   const styles = useStyles(getStyles);
   const [showingAddPanelModal, setShowingAddPanelModal] = useState(false);
 
   const useLibraryPanel = (panelInfo: LibraryPanelDTO) => {
+    const panelTypeChanged = panel.type !== panelInfo.model.type;
     panel.restoreModel({
-      ...panelInfo.model,
+      ...omit(panelInfo.model, 'type'),
       ...pick(panel, 'gridPos', 'id'),
       libraryPanel: toPanelModelLibraryPanel(panelInfo),
     });
 
+    if (panelTypeChanged) {
+      changePanelPlugin(panel, panelInfo.model.type);
+    }
+
     // Though the panel model has changed, since we're switching to an existing
     // library panel, we reset the "hasChanged" state.
     panel.hasChanged = false;
-
     panel.refresh();
     panel.events.publish(PanelQueriesChangedEvent);
+
+    // onChange is called here to force the panel editor to re-render
+    onChange();
```

Comment:
Actually, now that I'm trying this again, despite this calling `onPanelConfigChanged` and then `this.forceUpdate()` in the `PanelEditor`, it doesn't actually seem to re-render the panel editor, and I'm not sure why. Wonder if anyone has any ideas?

Reply:
Hmm, are you still struggling with this one? I think it works perfectly now 🎉